### DISCUSSION TITLE
Discussion: internal aliases

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#259] [General]
Ported from https://github.com/dotnet/roslyn/issues/7451.

While using-alias-directive "introduces an identifier that serves as an alias for a namespace or type within the immediately enclosing compilation unit or namespace body", internal aliases can be referenced anywhere within a program.
```cs
internal using StringMap = Dictionary<string , string>;
```
- Internal aliases will be erased during compile-time and will be replaced with the underlying type.
- Internal aliases are totally equivalent to the underlying type but, for example, IntelliSense could be extended to suggest only extension methods on the specific type alias.
- Internal aliases (along with file-scoped aliases) could also support generics (https://github.com/dotnet/csharplang/issues/90),

```cs
internal using Map<T> = Dictionary<string , T>;
```

Alternatively we could use `alias` or `type`, since `internal` carries some implications:
-  `InternalsVisibleTo` should work with internal aliases but it won't.
-  `internal` is "the default" modifier in a global or namespace context.

However, with CLR support we could extend the set of allowed modifiers for a type alias in the future.

EDIT: Following `using static` perhaps `using internal` is a better choice here.

**Comment by HaloFour** [comment]
So just to be clear, I could add a source file with the following contents:

```cs
namespace Foo {
    internal using Map<T> = System.Collections.Generic.Dictionary<string, T>;
}
```

And have some other source file with the following contents:

```cs
using Foo;

namespace Bar {
    static class Program {
        static void Main() {
            var map = new Map<int>();
            if (map is Dictionary<string, int>) Console.WriteLine("yay!");
        }
    }
}
```

And the compiler will resolve the `Foo.Map<T>` alias "pseudotype" to the actual type `System.Collections.Generic.Dictionary<string, T>`?

Also note that I used the fully-qualified type name to define the "alias" `Map<T>` as currently that is a requirement.  You don't appear to be proposing to change that.

**Reply by Shadowblitz16** [reply]
wouldn't it technically be public using?
internal is scoped to the assembly only.
What if I want to type alias something from what my library is built on?

**Comment by jnm2** [comment]
Whatever we do here, I hope we don't block a future CLR improvement where I can declare a runtime type alias, enabling me to change the name of a type without breaking source or binary compatibility.

**Comment by alrz** [comment]
@HaloFour 

>Also note that I used the fully-qualified type name to define the "alias" Map<T> as currently that is a requirement. You don't appear to be proposing to change that.

Here's the context.

```cs
using System.Collections.Generic;
namespace Foo {
    internal using Map<T> = Dictionary<string, T>;
}
```

To make that work in the same level, RHS should consider preceding using directives,
```cs
using System.Collections.Generic;
internal using Map<T> = Dictionary<string, T>;
````
Which might be a breaking change (not if we only do that for internal aliases).

**Comment by alrz** [comment]
@jnm2

>Whatever we do here, I hope we don't block a future CLR improvement where I can declare a runtime type alias, enabling me to change the name of a type without breaking source or binary compatibility.

Once CLR support this, we can extend the set of allowed modifiers to `public` and you'd get that in result.

**Comment by HaloFour** [comment]
@alrz 

IIRC, currently the compiler requires fully-qualified type names in aliases in order to avoid potential circular references.  How would you propose avoiding that issue with "internal aliases"?

**Comment by alrz** [comment]
@HaloFour

Could you give an example? Supposedly the compiler can detect such cases as it does for base classes.

**Comment by gulshan** [comment]
I think rather than redefining what `using` does, a new `typealias` or `typedef` or simply `type` should be use for this aliasing. And the full qualified name should not be required. And then this should have option to be `public`, usable from outside of the assembly.

**Comment by temporaryfile** [comment]
> Whatever we do here, I hope we don't block a future CLR improvement where I can declare a runtime type alias, enabling me to change the name of a type without breaking source or binary compatibility.

[Type Forwarding in the Common Language Runtime](https://msdn.microsoft.com/en-us/library/ms404275%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396)

**Comment by jnm2** [comment]
@playsomethingsaxman That allows moving a type between assemblies, which I have done, but there is no facility to typeforward a name change or namespace change.

**Comment by scottdorman** [comment]
> EDIT: Following `using static` perhaps `using internal` is a better choice here.

Wouldn't `using alias` actually be better? It's clearer that you're creating a type alias and doesn't carry forward the other implications that `internal` has (as you already pointed out).

**Comment by gulshan** [comment]
I think type parameters and user defined access modifiers look better with explicit aliasing syntax using some new keyword like `typealias`-

```csharp
public typealias StringKeyDictionary<T> = Dictionary<string, T>;
```

And `using` can remain scoped to files. If we prefer to introduce a keyword which can be used for other types of type definition, `typedef` of just `type` can be used, as I have already suggested.

**Comment by daiplusplus** [comment]
I'd like this to have an additional `strict` modifier that declares a type-alias as having the same binary/object representation as the underlying type but that it represents an incompatible value as far as business-rules are concerned - for example, this would help prevent mixing identity-types in database systems (which tend to all be `Int32`) without needing to define multiple new `struct`  types, for example:

    using strict CustomerId = Int32;
    using strict OrderId    = Int32;

    #region ORM/scaffold-generated code for business entities:

    class Order {
        public OrderId Id { get; set; }
        // ...
    }

    class Customer {
        public CustomerId Id { get; set; }
    }

    class Orm {
        public Order GetOrder(OrderId id) { ... }
        public Customer GetCustomer(CustomerId id) { ... }
    }

    #endregion

My proposed `strict` aliases are assignable to and from their underlying types without any problems, this is useful when passing values to other methods without casting, or when initializing values from a literal or other source:

    OrderId orderId = 123; // ok to assign from Int32 literal
    OrderId orderId = sqlDataReader.GetInt32(0); // ok to assign from Int32 return value

    Dictionary<Int32,Order> ordersById = ...
    Order order = ordersById[ orderId ]; // ok to use as Int32 argument value

Similarly:

    Customer customer = orm.GetCustomer( 123 );
    Order order       = orm.GetOrder( orderId: 456 );

While this code would cause a compile error:

    customer.Id = order.Id; // CustomerId and OrderId are not assignable
    Order order = orm.GetOrder( customer.Id ); // `CustomerId` values cannot be used as arguments for `OrderId` parameters.

This example of a database system is contrived and certainly a case where this isn't really a problem in practice, but it would be useful for zero-cost dimensional analysis.

`using strict` definitions could also be used to restrict extension methods, consider:

    using strict Rectangle<T> = Tuple<T,T>;

    public static T GetArea<T>(this Rectangle<T> rect) where T : INumber {
        return rect.Item1 * rect.Item2;
    }

    public static Single GetAreaInCentimetres(this Rectangle<Metre> rect) {
        return ( rect.Item1 * 100 ) * ( rect.Item2 * 100 );
    }

Which could then be used like so for zero-cost dimensional unit constraints:

    using strict Metre = Single;
    using strict Centimetre = Single;
    using strict MetricRectangle = Rectangle<Metre>;
    using strict Inch = Single;
    using strict USCustomaryRectangle = Rectangle<Inch>;

So this works:

    Rectangle<Metre> metricRect = new MetricRectangle( 0.5, 2.0 );
    Metre metricArea = metricRect.GetArea();

But this won't:

    Tuple<Single,Single> weakRect = new Tuple<Single,Single>( 0.5, 2.0 );
    Single weakArea = weakRect.GetArea();

Neither will this:

    USCustomaryRectangle inchesRect = new USCustomaryRectangle( 19.685, 78.7402 );
    Centimetre cm = inchesRect.GetAreaInCentimetres();

**Comment by osdm** [comment]
How about allowing this inside generic classes? For example:

    public class MyOneToListMap<TKey, TValue>
    {
        using TValueList = List<TValue>;
        using TDictionary = Dictionary<TKey, TValueList>;
        private readonly TDictionary dictionary = new TDictionary();
    }

P.S. Although in my example you can just make derived classes, I've encountered many cases where derived classes are impossible (for example, you must accept/return externally declared interfaces with many type parameters).

**Comment by pongba** [comment]
> Whatever we do here, I hope we don't block a future CLR improvement where I can declare a runtime type alias, enabling me to change the name of a type without breaking source or binary compatibility.

Agreed, exactly what I need

**Comment by john-h-k** [comment]
You could emit metadata into the assembly, some sort of attribute which defines the aliases (type, string). Then `InternalsVisibleTo` could work, right?

```[assembly: InternalAlias(typeof(Dictionary<string, string>), "StringMap")]``` etc

**Comment by wfowler1** [comment]
I believe this is related to my situation. I use a lot of preprocessor definitions in my projects to conditionally compile aliases into my code. For example, if I have a piece of code designed for use in .NET standalone as well as Unity3D engine:

```
class MyClass
{
#if UNITY3D
	using Vector3 = UnityEngine.Vector3;
#else
	using Vector3 = System.Numerics.Vector3;
#endif
	// ...
}
```
Currently this snippet needs to be copied across every file that needs to use a Vector3 instance. What I'd really like is to be able to declare an alias like this across an entire namespace. Since as far as I know I don't have control over the assembly metadata Unity3D produces, this is currently the best way to do this.

**Comment by john-h-k** [comment]
A complete rework of using aliases would be nice:

* Allow them to be used across an assembly

* Store them as metadata so they can be used from other assemblies optionally, and can be decompiled properly 

* Allow using C# primitive aliases like `int`

* Allow using pointers

* Allow using `ref`s?? Maybe

**Comment by CyrusNajmabadi** [comment]
@wfowler1 You could define your own struct that conditionally wraps either of these.  Then just reference your struct everywhere.

**Comment by daiplusplus** [comment]
@CyrusNajmabadi 

As C# doesn't support struct inheritance we can't simply subclass the original types with zero-overhead:

```
public struct MyVector3 :
#if UNITY
    Unity.Vector3
#else
    Numerics.Vector3
#endif
{
}
```

Now, assuming you're referring to wrapping it like this:

```
struct UnityVector3
{
	public Single X;
	public Single Y;
	public Single Z;
}

struct SystemVector3
{
	public Single X;
	public Single Y;
	public Single Z;
}

struct Vector3Wrapper
{
#if UNITY
	private UnityVector3 Value;
#else
	private SystemVector3 Value;
#endif

	public Single X
	{
		get { return this.Value.X; }
		set { this.Value.X = value; }
	}

	public Single Y
	{
		get { return this.Value.Y; }
		set { this.Value.Y = value; }
	}

	public Single Z
	{
		get { return this.Value.Z; }
		set { this.Value.Z = value; }
	}
}
```

There are at least a few issues with this approach:

1. Structs can expose raw fields, which means you can pass values directly in `ref` contexts and take the address of a field in `unsafe` contexts and the semantics of certain expressions (such as those involving the `++` operator) are guaranteed to reduce to the simplest string of CPU instructions, but having to go via properties means you lose these advantages.
2. I disagree about it being a zero-cost abstraction: while the CLR's memory layout of value-types is not well documented (by design), I don't believe it is documented anywhere that a `struct` that has a single struct field and no other fields has identical memory layout compared to the nested struct. Additionally, there are no guarantees that the JIT will inline or optimize-away trivial proxy properties.
3. Ergonomically it's a problem because now, for each member of each struct you now need to define those trivial proxy properties. Until C# gets true duck-typing and structual typing support this is something you would probably have to do in T4 to save on typing.
4. When/if interfacing with the original types (e.g. when exchanging data with the Unity game engine) you need to pass the wrapped value or perform a conversion. As structs are passed by-value this means an extra copy for each struct instance passed, which adds up if you're using a large `Vector3`-based object graph to represent a complex 3D model, for example.

**Comment by wfowler1** [comment]
Yes, the reasons @Jehoel cited make me shy away from that approach. Even using an approach which masks the issue by using implicit conversions is an ugly way to do things.

**Comment by CyrusNajmabadi** [comment]
THe jit should have no problems with any of that.  This is an exceedingly common pattern, and is precisely how you can have a zero-overhead abstraction over other types (useful when you want to hide functionality or add new functionality).

**Comment by john-h-k** [comment]
@Jehoel

> while the CLR's memory layout of value-types is not well documented (by design)

Yes, but given the fact the struct stores no other data, there is no reason for the CLR to make it any larger. It doesn't just add padding for fun

**Comment by faldor20** [comment]
Any idea if any progress is going to be made on this? It would seem very simple to implement this as just replacing what you wrote with the full type name during compilation. Just like a #define works in c or c++.
I use this type a fair bit in some recent code: 
Dictionary<MailboxAddress, Dictionary<NotificationOutput, AssemblyAction[]>>
I despise writing it out each time.

**Comment by CyrusNajmabadi** [comment]
> Any idea if any progress is going to be made on this?

There is currently no champion for this item, so the ldm is not prioritizing this in any way. 

> It would seem very simple to implement this as just replacing what you wrote with the full type name during compilation.

It's definitely not that simple :-)

**Comment by alrz** [comment]
Some things to consider in this space:

- generic type aliases 
- primitive type aliases (int, tuples, function pointers, etc)
- global namespace aliases

There are already proposals for the first two, but I think this is a better place to invest instead of extending `using` directive.

For namespaces this would be really impactful for layered applications, e.g. in a web app you may have a domain model, api model and business model all with the same name. Using extern alias for each is even less convinient than repeating `using` on top of each file which still sounds like a workaround and at the end it's really hard to manage all of those and keep them in sync.


F# already supports this (I think only for types) and it's speced that these won't be accessible from outside of the assembly and that haven't been much of a problem so far.

I suspect one reason for this to not get prioritized is that there's no immediate usecase in roslyn or corefx, but it's an absolute necessity in other kinds of applications.

**Comment by gulshan** [comment]
So proposals useful for Roslyn or CoreFx are prioritized? Maybe that's natural.

**Comment by YairHalberstadt** [comment]
@gulshan 
Without a customer within Microsoft it's very difficult to thrash out the problem domain, and work out what is actually sufficient to solve the problem. Otherwise you're creating features and hoping that they solve the problems people are having, but not finding out till after they're already released.

**Comment by 333fred** [comment]
> I suspect one reason for this to not get prioritized is that there's no immediate usecase in roslyn or corefx, but it's an absolute necessity in other kinds of applications.

I don't think this is true. Rather, I think that internal-only aliases is not the approach we want to take. For example, are these a compile-time only construct? If so, what happens to assemblies with IVT? The only logical answer is that they see the original type, not the alias. And if it's not a compile-time only construct, and somehow encoded in metadata, then it would be sad if you couldn't use these in a public interface. But then we have the question of whether aliases are identity-convertible to their underlying type. You might want that for a simple alias of a complicated generic type, but if you were creating a units library, for example, you might want your miles alias to not be identity-convertible to your kilometers alias. Plus, a metadata encoding would likely want runtime support, particularly for the opaque version. So, I don't think that the problem is not something that we want to solve. Rather, I think that we haven't found a proposal yet that we like, and no one on the ldt has sat down to make one either.

**Comment by alrz** [comment]
FWIW I think the introduction of DUs would make another use case for global aliases, for instance you may have a generic Result<T, E> and you'd want to define aliases for each kind of error:

`type ServiceResult<T> = Result<T, ServiceError>`

I think that is how all Rust's result types are defined.

**Reply by softlion** [reply]
That is exactly what i wanted to do today:

`typedef UnitVector3 = Vector3`

So my extension methods already knows the vector is normalized.  
Like in MathNet, but much better architectured.

I have no workaround except embedding a Vector3 in a new class, as Vector3 is sealed.

**Comment by jukkahyv** [comment]
This has been my personal favorite feature request for a while. It would increase our team's productivity significantly, and we can't be the only ones. I already explained our case at Build Developer Day, but let me post it there too.

Our system has many, many entities (over 200). It's common that their IDs are passed around as plain integers. There's been many cases where developer mixes IDs of different, similar sounding entities, let's say UserContact and UserConfig. So I got an idea to make strongly typed IDs for entities, to avoid mixing them. It's implemented as struct ```EntityID<TEntity>```, for example ```EntityID<UserContact>```. But writing ```EntityID<UserContact>``` everywhere is tedious, so we would like to have a type UserContactID that is compatible with ```EntityID<UserContact>```. Obviously structs cannot be inherited, but even with classes, UserContactID and ```EntityID<UserContact>``` would be two different types: I cannot have a method that expects UserContactID and is given ```EntityID<UserContact>``` as parameter, even though these types are identical. I could work around this with implicit casts, but it's hacky at best, and doesn't work nearly everywhere (what if I have a collection of UserContactIDs?). Many other languages such as C++ have type aliases which would solve this use case. From what I remember from my C++ days, typedef was exactly what would be needed here. It seems simple, but I understand there are corner cases. Ideally we would want these to be public so that they can be shared between assemblies, but even internal aliases would be a big help.

**Comment by gulshan** [comment]
@jukkahyv Let me just put a plug for my proposal #3046 here, which could be helpful for your specific case. It purposes to have generic types defined as `T.EntityId<T>` and then use as `UserContact.EntityId` (or shorter- define as `T.Id<T>`, use as `UserContact.Id`).

**Comment by Happypig375** [comment]
Plugging an already negatively-reviewed proposal probably isn't a good idea.

Regarding the actual relevance of the proposal, `UserContact.Id` expands to `Id<UserContact>` which is not the original type as defined by @jukkahyv (`EntityID<UserContact>`). If you were to write `UserContact.EntityID`, then it is just as long as `EntityID<UserContact>`.

**Comment by jukkahyv** [comment]
Actually, I forgot to mention one big point. By making it a single name (like UserContactId), it can be autocompleted with intellisense much faster than generic type + parameter. It sounds like a small thing, and it is, until you realize you're doing it 10 or 50 times per day.

**Comment by Happypig375** [comment]
And here I use F# type inference every day. 😃

**Comment by declard** [comment]
Imagine C# had aliases allowing this:
```csharp
public abstract class Base
{
  public abstract alias Inner<T>;
}

public Generic<T> {}
public Impl : Base
{
  public override alias Inner<T> = Generic<T>;
}
```
Then HKT would become unnecessary as you could be able to emulate them using this technique:
```csharp
H<T> F<H<>, T>(H<T> h) => h;
```
becomes
```csharp
class H
{
  public abstract alias Inner<T>;
}
G.Inner<T> F<G, T>(G.Inner<T> h) where G : H => h;
```

**Reply by softlion** [reply]
What is HKT

**Reply by Joe4evr** [reply]
Higher Kinded Type. See #339 for a more in-depth proposal.

**Comment by VBAndCs** [comment]
This will be more powerful if also allows generic aliases #3309.

**Comment by ChayimFriedman2** [comment]
@Jehoel 

I greatly agree with your suggestion, but I think it's subject for a different proposal.

I would define `internal using` (or `using internal`) as C++'s `typedef`: type alias - shorten a type name.

For your proposal, I would suggest `type T = `, as you're creating a new type.

Also, this proposal don't require CLR changes (although can benefit from them) - even public aliases could be exposed as attributes, as already pointed out. But your proposal **requires** CLR changes, because otherwise the following code will result in wrong:
```c#
type MyType = int;

// ...

object o = new MyType(5);
if (o is int)
{
    Console.WriteLine("Oops, they should be different types.");
}
```

Although now it's crossing my mind that it can be implemented by the compiler by new struct containing the type, and re-exposing all of its public members:

```c#
struct MyType
{
    int _Field;
    // All int members
}
```

This **will** have overhead, but since the JIT will optimize this away this should be fine.

**Reply by CleanCodeX** [reply]
What is the progress on this topic?

**Reply by alrz** [reply]
@CleanCodeX See https://github.com/dotnet/csharplang/issues/3428

**Comment by YoshiRulz** [comment]
Does there exist a hack or third-party solution for the simplest form of this: a global (per compilation unit) alias for a non-generic type, which only exists at compile-time? I can't find anything that's not hypothetical in this thread, in #7451, or on the web.

**Reply by daiplusplus** [reply]
@YoshiRulz The only feasible option right now is a Roslyn code-generator that creates an almost-zero-overhead `struct` with implicit conversion defined to/from types you want it to be compatible with.

Note that Roslyn code-generators can't modify existing code on-disk, they can only add new code in hidden `.cs` files that exist only in-memory (though they can be saved to your `obj` folder during debugging). So I guess you'd define a custom Attribute class in your project to declare these "aliases" and have your Roslyn code-generator recognize them and generate the appropriate type at build-time (VS will run Roslyn code-generators at design-time too).

I can't see anything on NuGet that does this right now, but it is fairly straightforward, so I'm surprised no-one's done it already.
